qiskit/primitives/statevector_estimator.py

```diff
@@ -0,0 +1,65 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023, 2024.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Estimator class
+"""
+
+from __future__ import annotations
+
+from collections.abc import Iterable
+
+import numpy as np
+
+from qiskit.quantum_info import SparsePauliOp, Statevector
+
+from .base import BaseEstimatorV2
+from .containers import EstimatorPub, EstimatorPubLike, PrimitiveResult, PubResult
+from .primitive_job import PrimitiveJob
+from .utils import bound_circuit_to_instruction
+
+
+class StatevectorEstimator(BaseEstimatorV2):
+    """
+    Simple implementation of :class:`BaseEstimatorV2` with full state vector simulation.
+    """
+
+    def run(
+        self, pubs: Iterable[EstimatorPubLike], *, precision: float | None = None
+    ) -> PrimitiveJob[PrimitiveResult[PubResult]]:
+        if precision is not None and precision > 0:
+            raise ValueError("precision must be None or 0 for StatevectorEstimator.")
```

Comment:
```suggestion
        if precision is None:
            precision = 0
```
We don't want to error if a user puts in `precision > 0` value for precision (which is a target precision), since we will always exceed it since precision is guaranteed to be 0 by implementation. This would prevent actually being able to use this in any application / algorithm workflow which uses a sensible initial guess for precision which would always be > 0.

Reply:
@chriseclectic do you propose to:
1) ignore the precision
2) artificially add imprecision, as was previously implemented

The argument for (1) is that this is a statevector simulation, so it should always be perfect.
The argument for (2) is that some folks may want to use this implementation to test their ideas, which might include testing that their ideas are robust to finite noise.

Reply:
1: in the base class we say that it is a target precision, which i implicitly assumed meant "I want you to return me a mean estimate with standard error <= this value". Maybe we can make wording more explicit.

To add to my lack of documentation comments, this classes doc string should explain how the simulation is done and say that it will always return standard error as zero because it returns exact expectation value means computed from the full state vector.

Reply:
But (1) and (2) both satisfy "return me an estimate with standard error <= this value". Sure, (1) does a better estimation job, but 2 technically satisfies the requirements if the sampling is done right.

Reply:
Sure, but neither raises an error and makes the estimator unusable in portable workflow like this does

Reply:
Also i dont see the point in artificially sampling noise into a perfect estimate for a reference implementation, which is what this is. You can always do that from the results if you really want (and wrap it as another estimator if you like if its for testing for example)

Reply:
For users who set precision>0 expecting to run simulations with shot noise, it is difficult to understand why the implementation automatically sets it to 0. You say there are no such users, but they exist, and they are using this feature.

Reply:
I agree with @ikkoham : the main point of the `StatevectorEstimator` is not really to compute expectation values for users, it is to provide a simple reference implementation, and a tool for testing. This is because there are other implementations that are strictly better at computing expectation values: for non-exponential (albeit nisqy) runtime you need to use a runtime backend, and for more performant classical simulations (including Clifford) you can use aer. I realize this is a reversal of some previous comments I made, but I've had a bit more time to think about it now.

So, I am in favour of `(2) artificially add imprecision, as was previously implemented` because `some folks may want to use this implementation to test their ideas, which might include testing that their ideas are robust to finite noise`. If some user is in the unlikely situation that they want to call `some_application_funcion(estimator: BaseEstimatorV2)` with noiseless estimator, and they know that `some_application_function` internally hard-codes specific precisions >0, they can subclass `StatevectorEstimator`:

```python
class NoiselessEstimator(StatevectorEstimator):
    def run(self, pubs, precision = None):
        return super().run(pubs, 0.0)

result = some_application_function(NoiselessEstimator())
```

I do think that the default precision for `StatevectorEstimator` should be 0, however.

Reply:
I don't see why a user should expect this estimator to be noisy unless we tell them that it is. I think we should explicitly state this is an ideal estimator and the precision argument is ignored since it always perfectly estimates the observable in the class documentation. 

You need to consider how an estimator is used in computational workflows. If someone is writing an algorithm (say VQE for example) that takes an estimator to run on,  it will choose some initial guess for a precision, and then perhaps depending on the variance of the result may modify this processing for subsequent evaluations. If you have some special test Estimator that only works with an argument of precision 0 this will error in these sort of applications. You don't loose anything by return a precision less than requested. 

If you want to add noise to this estimator so precision effects the results artificially you can do so (or make a second subclass estimator that adds noise), but i will not approve this PR while it raises an error or warning under a canonical workflow.

Reply:
(Also, to be clear, and I think we all three agree on this point: the `StatevectorEstimator` shouldn't raise errors when given `precision>0` in any case. That would make the thing non-portable, in addition to being annoying. So I'm in particular trying to get to the bottom of choosing between 1 and 2, outlined near the top of this thread)

Reply:
If you want the default to be to add noise I am fine with that, just dont add a shots option, all you shuold need is a `seed` field (like sampler) and initialize an rng in run and do something like `rng.normal(evs, precision)` (im not exactly sure correct syntax for normal distribution sampling in numpy)

Reply:
Great, let's go with (2) and implement rng seeding in the same way as the `StatevectorSampler`. No need to mention shots anywhere.